File: src/frontend/src/api/std.ts
import constants from "@/constants";

// backend.api.std.StdWattsParams
export interface StdWattsParams {
    std_threshold: number;
    data_exp_name: string;
    app_name: string;
}

// backend.api.std.StdWattsResponse
export interface StdWattsResponse {
    lower: number;
    upper: number;
    mean: number;
    std: number;
}

export async function getStdWatts(params: StdWattsParams): Promise<StdWattsResponse> {
    const resp = await fetch(`${constants.backendApiUrl}/std/watts`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(params),
    });
    return (await resp.json()) as StdWattsResponse;
}

File: src/backend/api/std.py
import joblib
import numpy as np
from fastapi import APIRouter
from pydantic import BaseModel

import enilm.etypes

from ..mem import cache_folder
from ..data import get_data

router = APIRouter(prefix="/std")


class StdWattsParams(BaseModel):
    std_threshold: float
    data_exp_name: str
    app_name: enilm.etypes.AppName


class StdWattsResponse(BaseModel):
    lower: float
    upper: float
    mean: float
    std: float


@router.post("/watts")
async def getStdWatts(params: StdWattsParams) -> StdWattsResponse:
    memory: joblib.Memory = joblib.Memory(cache_folder / "std_watts", verbose=1)

    @memory.cache
    def get_std_watts(params: StdWattsParams) -> StdWattsResponse:
        data = get_data(params.data_exp_name).resampled_data
        assert params.app_name in data
        data = data[params.app_name]

        mean_data = np.mean(data)
        std_data = np.std(data)
        return StdWattsResponse(
            lower=float(mean_data - params.std_threshold * std_data),
            upper=float(mean_data + params.std_threshold * std_data),
            mean=float(mean_data),
            std=float(std_data),
        )

    return get_std_watts(params)
